File: modules/benchmarks/arrow-scala-benchmarks/src/main/scala/cats/Pure.scala
package arrow.benchmarks.effects.scala.cats

import cats.effect.IO

object Pure {

  def ioPureLoop(size: Int, i: Int): IO[Int] =
    IO.pure(i).flatMap { j =>
      if (j > size) IO.pure(j) else ioPureLoop(size, j + 1)
    }

  def unsafeIOPureLoop(size: Int, i: Int): Int =
    ioPureLoop(size, i).unsafeRunSync()

}

File: modules/benchmarks/arrow-kio-benchmarks/src/main/kotlin/kio/Pure.kt
package arrow.benchmarks.effects.kio

import it.msec.kio.UIO
import it.msec.kio.flatMap
import it.msec.kio.just
import it.msec.kio.runtime.Runtime

object Pure {

  fun kioPureLoop(size: Int, i: Int): UIO<Int> =
    just(i).flatMap { j ->
      if (j > size) just(j) else kioPureLoop(size, j + 1)
    }

  fun unsafeKIOPureLoop(size: Int, i: Int): Int =
    Runtime.unsafeRunSyncAndGet(kioPureLoop(size, i))
}
